Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs

```diff
@@ -96,11 +109,79 @@ private static IEnumerable<MemberResult> GetScopeVariables(ModuleAnalysis analys
             return res;
         }
 
+        private DocumentSymbol[] ToDocumentSymbols(List<MemberResult> members) {
+            var childMap = new Dictionary<MemberResult, List<MemberResult>>();
+            var topLevel = new List<MemberResult>();
+
+            foreach (var m in members) {
+                var parent = members.FirstOrDefault(x => x.Scope?.Node == m.Scope?.OuterScope?.Node && x.Name == m.Scope?.Name);
+                if (parent != default(MemberResult)) {
+                    if (!childMap.TryGetValue(parent, out var children)) {
+                        childMap[parent] = children = new List<MemberResult>();
+                    }
+                    children.Add(m);
+                } else {
+                    topLevel.Add(m);
+                }
+            }
+
+            var symbols = topLevel
+                .GroupBy(mr => mr.Name)
+                .Select(g => g.First())
+                .Select(m => ToDocumentSymbol(m, childMap))
+                .ToArray();
+
+            return symbols;
+        }
+
+        private DocumentSymbol ToDocumentSymbol(MemberResult m, Dictionary<MemberResult, List<MemberResult>> childMap) {
+            var res = new DocumentSymbol {
+                name = m.Name,
+                detail = m.Name,
+                kind = ToSymbolKind(m.MemberType),
+                deprecated = false,
+                _functionKind = GetFunctionKind(m)
+            };
+
+            if (childMap.TryGetValue(m, out var children)) {
+                res.children = children.Select(x => ToDocumentSymbol(x, childMap)).ToArray();
+            } else {
+                res.children = new DocumentSymbol[0];
+            }
+
+            var loc = m.Locations.FirstOrDefault(l => !string.IsNullOrEmpty(l.FilePath));
+            if (loc != null) {
+                res.range = new SourceSpan(
+                        new SourceLocation(loc.StartLine, loc.StartColumn),
+                        new SourceLocation(loc.EndLine ?? loc.StartLine, loc.EndColumn ?? loc.StartColumn)
+                    );
+                res.selectionRange = res.range;
+            }
+            return res;
+        }
+
+        private static string GetFunctionKind(MemberResult m) {
+            if (m.MemberType == PythonMemberType.Function) {
+                var funcDef = m.Values.FirstOrDefault(x => x is IPythonFunction)?.PythonType as FunctionDefinition;
```

Comment:
This always results in `null` - you will want `m.Values.OfType<FunctionInfo>()` and then check their `.IsStatic`/`.IsClassMethod`/`.IsProperty` member.

Reply:
OK